File: mergesort.py
import random

def merge(ls1, ls2):
    '''Helper function for MergeSort
    This function assumes both source arrays are sorted,
    then builds a result array by comparing the first
    elements from each and storing the lower one'''
    result = []
    while ls1 and ls2:
        if ls1[0] < ls2[0]:
            result.append(ls1.pop(0))
        else:
            result.append(ls2.pop(0))
    return result + ls1 + ls2

def MergeSort(ls):
    '''Sorts the input list using a recursiv
    mergesort algorithm'''
    if len(ls) > 1:
        mid = len(ls) // 2
        return merge(MergeSort(ls[:mid]),MergeSort(ls[mid:]))
    else:
        return ls



#print(merge([3,5,7,9],[2,4,6,8,10,12]))    
print(MergeSort([7,13,2,9,17,4,6,1]))
print(MergeSort([]))

random_list = []
for _ in range(1000):
    random_list.append(int(random.random()*1000))

print(MergeSort(random_list))

File: mergesort.js
// Helper function for MergeSort
// This function assumes both source arrays are sorted,
// then builds a result array by comparing the first
// elements from each and storing the lower one
const merge = (arr1, arr2) => {
    result = [];
    while (arr1.length && arr2.length) {
        (arr1[0] < arr2[0]) ? result.push(arr1.shift()) : result.push(arr2.shift())
    }
    return result.concat(arr1).concat(arr2)
}

// Sorts the input list using a recursiv
// mergesort algorithm
const MergeSort = arr => {
    if (arr.length > 1) {
        let mid = Math.floor(arr.length / 2);
        return merge(MergeSort(arr.slice(0,mid)),MergeSort(arr.slice(mid,mid.length)));
    }
    return arr;
}

//console.log(merge([2,4,6],[3,5,7,9]))

random_list = []
for (let i = 0; i < 100; i++) {
    random_list.push(Math.floor(Math.random()*100));
}
console.log(MergeSort(random_list));